File: experiment/websites/c/gnomesort/gnomesort.c
/* based on: https://rosettacode.org/wiki/Sorting_algorithms/Gnome_sort#C (2021-12-23) */

#include <emscripten/emscripten.h>

EMSCRIPTEN_KEEPALIVE
void gnomesort(int *a, int n)
{
  int i=1, j=2, t;
# define swap(i, j) { t = a[i]; a[i] = a[j]; a[j] = t; }
  while(i < n) {
    if (a[i - 1] > a[i]) {
      swap(i - 1, i);
      if (--i) continue;
    }
    i = j++;
  }
# undef swap
}

File: experiment/websites/js/gnomesort/gnomesort.js
/* based on: https://rosettacode.org/wiki/Sorting_algorithms/Gnome_sort#JavaScript (2021-12-23) */

/**
 * @param {number[]} a
 * @returns {number[]}
 */
function gnomesort(a) {
  function moveBack(i) {
    for( ; i > 0 && a[i-1] > a[i]; i--) {
      var t = a[i];
      a[i] = a[i-1];
      a[i-1] = t;
    }
  }
  for (var i = 1; i < a.length; i++) {
    if (a[i-1] > a[i]) moveBack(i);
  }
  return a;
}
